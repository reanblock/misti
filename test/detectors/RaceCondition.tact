// Test cases for RaceCondition detector

// TEST 1: Classic race condition with query_id
contract QueryIdRace {
    query_id: Int = 0;
    
    // BAD: Writes query_id and sends message
    receive(msg: Request) {
        self.query_id = msg.query_id;  // Write state
        send(SendParameters{
            to: msg.target,
            value: ton("0.1"),
            body: ForwardRequest{query_id: self.query_id}.toCell()
        });
        // Another message can arrive here and change query_id!
    }
    
    // BAD: Reads query_id that could have been modified
    receive(msg: Response) {
        if (msg.query_id != self.query_id) {  // Read potentially stale state
            throw(123);
        }
    }
}

// TEST 2: Multiple receivers accessing same state
contract SharedStateRace {
    counter: Int = 0;
    lastSender: Address;
    
    // BAD: Writes counter and sends
    receive(msg: Increment) {
        self.counter = self.counter + 1;
        self.lastSender = sender();
        send(SendParameters{
            to: self.owner,
            value: ton("0.05"),
            body: CounterUpdated{value: self.counter}.toCell()
        });
    }
    
    // BAD: Reads counter that increment might be modifying
    receive(msg: GetCounter) {
        self.reply(CounterValue{value: self.counter}.toCell());
    }
    
    // BAD: Also accesses the shared counter
    receive(msg: Reset) {
        self.counter = 0;
    }
}

// TEST 3: State mutation in functions called by receivers
contract IndirectRace {
    nonce: Int = 0;
    
    // BAD: Function that mutates and sends
    fun processRequest() {
        self.nonce = self.nonce + 1;
        send(SendParameters{
            to: self.processor,
            value: ton("0.1"),
            body: ProcessNonce{nonce: self.nonce}.toCell()
        });
    }
    
    receive(msg: StartProcess) {
        self.processRequest();
    }
    
    // BAD: Reads nonce that processRequest might be modifying
    receive(msg: VerifyNonce) {
        require(msg.nonce == self.nonce, "Invalid nonce");
    }
}

// TEST 4: Good pattern - no race condition
contract NoRaceGoodPattern {
    totalProcessed: Int = 0;
    
    // GOOD: Uses local variable for consistency
    receive(msg: Process) {
        let localQueryId = msg.query_id;  // Use local variable
        send(SendParameters{
            to: msg.target,
            value: ton("0.1"),
            body: ProcessRequest{query_id: localQueryId}.toCell()
        });
        self.totalProcessed = self.totalProcessed + 1;  // Safe counter update
    }
    
    // GOOD: Only reads immutable or safely updated state
    get fun getTotalProcessed(): Int {
        return self.totalProcessed;
    }
}

// TEST 5: Complex race with multiple state variables
contract ComplexRace {
    stage: Int = 0;
    participant: Address;
    amount: Int = 0;
    
    // BAD: Multiple state mutations with send
    receive(msg: InitiateTransfer) {
        self.stage = 1;
        self.participant = msg.target;
        self.amount = msg.amount;
        
        send(SendParameters{
            to: self.participant,
            value: ton("0.1"),
            body: TransferInit{
                stage: self.stage,
                amount: self.amount
            }.toCell()
        });
    }
    
    // BAD: Reads multiple potentially stale values
    receive(msg: ConfirmTransfer) {
        require(self.stage == 1, "Wrong stage");
        require(msg.participant == self.participant, "Wrong participant");
        require(msg.amount == self.amount, "Wrong amount");
        
        self.stage = 2;
    }
}

// Required message types for tests
message Request {
    query_id: Int as uint64;
    target: Address;
}

message Response {
    query_id: Int as uint64;
}

message ForwardRequest {
    query_id: Int as uint64;
}

message Increment {}

message GetCounter {}

message Reset {}

message CounterUpdated {
    value: Int as uint32;
}

message CounterValue {
    value: Int as uint32;
}

message StartProcess {}

message ProcessNonce {
    nonce: Int as uint64;
}

message VerifyNonce {
    nonce: Int as uint64;
}

message Process {
    query_id: Int as uint64;
    target: Address;
}

message ProcessRequest {
    query_id: Int as uint64;
}

message InitiateTransfer {
    target: Address;
    amount: Int as coins;
}

message TransferInit {
    stage: Int as uint8;
    amount: Int as coins;
}

message ConfirmTransfer {
    participant: Address;
    amount: Int as coins;
}