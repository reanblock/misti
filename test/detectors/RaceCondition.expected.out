[HIGH] RaceCondition: Potential race condition: Function "QueryIdRace::receive(Request)" writes to state variable "query_id" and sends messages. This variable is also accessed by: QueryIdRace::receive(Response). Other messages may arrive and modify this state between send and response.
test/detectors/RaceCondition.tact:8:5:
  7 | 
  8 |     receive(msg: Request) {
    |     ^
Help: Consider using local variables or message-specific storage instead of shared state for values that must remain consistent across async message flows.

[HIGH] RaceCondition: Potential race condition: Receiver "QueryIdRace::receive(Response)" reads state variable "query_id" that is written by message-sending functions: QueryIdRace::receive(Request). The value may change between when it was set and when this receiver reads it.
test/detectors/RaceCondition.tact:19:5:
 18 | 
 19 |     receive(msg: Response) {
    |     ^
Help: Pass necessary data as part of the message instead of relying on shared state that can change.
This pattern is particularly dangerous with query_id or nonce values that must match between request and response.

[HIGH] RaceCondition: Potential race condition: Function "SharedStateRace::receive(Increment)" writes to state variable "counter" and sends messages. This variable is also accessed by: SharedStateRace::receive(GetCounter), SharedStateRace::receive(Reset). Other messages may arrive and modify this state between send and response.
test/detectors/RaceCondition.tact:32:5:
 31 | 
 32 |     receive(msg: Increment) {
    |     ^
Help: Consider using local variables or message-specific storage instead of shared state for values that must remain consistent across async message flows.

[HIGH] RaceCondition: Potential race condition: Function "SharedStateRace::receive(Increment)" writes to state variable "lastSender" and sends messages. This variable is also accessed by: SharedStateRace::receive(Increment). Other messages may arrive and modify this state between send and response.
test/detectors/RaceCondition.tact:32:5:
 31 | 
 32 |     receive(msg: Increment) {
    |     ^
Help: Consider using local variables or message-specific storage instead of shared state for values that must remain consistent across async message flows.

[HIGH] RaceCondition: Potential race condition: Receiver "SharedStateRace::receive(GetCounter)" reads state variable "counter" that is written by message-sending functions: SharedStateRace::receive(Increment). The value may change between when it was set and when this receiver reads it.
test/detectors/RaceCondition.tact:43:5:
 42 | 
 43 |     receive(msg: GetCounter) {
    |     ^
Help: Pass necessary data as part of the message instead of relying on shared state that can change.
This pattern is particularly dangerous with query_id or nonce values that must match between request and response.

[HIGH] RaceCondition: Potential race condition: Function "IndirectRace::processRequest" writes to state variable "nonce" and sends messages. This variable is also accessed by: IndirectRace::receive(VerifyNonce), IndirectRace::processRequest. Other messages may arrive and modify this state between send and response.
test/detectors/RaceCondition.tact:58:5:
 57 | 
 58 |     fun processRequest() {
    |     ^
Help: Consider using local variables or message-specific storage instead of shared state for values that must remain consistent across async message flows.

[HIGH] RaceCondition: Potential race condition: Receiver "IndirectRace::receive(VerifyNonce)" reads state variable "nonce" that is written by message-sending functions: IndirectRace::processRequest. The value may change between when it was set and when this receiver reads it.
test/detectors/RaceCondition.tact:71:5:
 70 | 
 71 |     receive(msg: VerifyNonce) {
    |     ^
Help: Pass necessary data as part of the message instead of relying on shared state that can change.
This pattern is particularly dangerous with query_id or nonce values that must match between request and response.

[HIGH] RaceCondition: Potential race condition: Function "ComplexRace::receive(InitiateTransfer)" writes to state variable "stage" and sends messages. This variable is also accessed by: ComplexRace::receive(ConfirmTransfer), ComplexRace::receive(InitiateTransfer). Other messages may arrive and modify this state between send and response.
test/detectors/RaceCondition.tact:103:5:
102 | 
103 |     receive(msg: InitiateTransfer) {
    |     ^
Help: Consider using local variables or message-specific storage instead of shared state for values that must remain consistent across async message flows.

[HIGH] RaceCondition: Potential race condition: Function "ComplexRace::receive(InitiateTransfer)" writes to state variable "participant" and sends messages. This variable is also accessed by: ComplexRace::receive(ConfirmTransfer), ComplexRace::receive(InitiateTransfer). Other messages may arrive and modify this state between send and response.
test/detectors/RaceCondition.tact:103:5:
102 | 
103 |     receive(msg: InitiateTransfer) {
    |     ^
Help: Consider using local variables or message-specific storage instead of shared state for values that must remain consistent across async message flows.

[HIGH] RaceCondition: Potential race condition: Function "ComplexRace::receive(InitiateTransfer)" writes to state variable "amount" and sends messages. This variable is also accessed by: ComplexRace::receive(ConfirmTransfer), ComplexRace::receive(InitiateTransfer). Other messages may arrive and modify this state between send and response.
test/detectors/RaceCondition.tact:103:5:
102 | 
103 |     receive(msg: InitiateTransfer) {
    |     ^
Help: Consider using local variables or message-specific storage instead of shared state for values that must remain consistent across async message flows.

[HIGH] RaceCondition: Potential race condition: Receiver "ComplexRace::receive(ConfirmTransfer)" reads state variable "stage" that is written by message-sending functions: ComplexRace::receive(InitiateTransfer). The value may change between when it was set and when this receiver reads it.
test/detectors/RaceCondition.tact:119:5:
118 | 
119 |     receive(msg: ConfirmTransfer) {
    |     ^
Help: Pass necessary data as part of the message instead of relying on shared state that can change.
This pattern is particularly dangerous with query_id or nonce values that must match between request and response.

[HIGH] RaceCondition: Potential race condition: Receiver "ComplexRace::receive(ConfirmTransfer)" reads state variable "participant" that is written by message-sending functions: ComplexRace::receive(InitiateTransfer). The value may change between when it was set and when this receiver reads it.
test/detectors/RaceCondition.tact:119:5:
118 | 
119 |     receive(msg: ConfirmTransfer) {
    |     ^
Help: Pass necessary data as part of the message instead of relying on shared state that can change.
This pattern is particularly dangerous with query_id or nonce values that must match between request and response.

[HIGH] RaceCondition: Potential race condition: Receiver "ComplexRace::receive(ConfirmTransfer)" reads state variable "amount" that is written by message-sending functions: ComplexRace::receive(InitiateTransfer). The value may change between when it was set and when this receiver reads it.
test/detectors/RaceCondition.tact:119:5:
118 | 
119 |     receive(msg: ConfirmTransfer) {
    |     ^
Help: Pass necessary data as part of the message instead of relying on shared state that can change.
This pattern is particularly dangerous with query_id or nonce values that must match between request and response.